etna/datasets/tsdataset.py

```diff
@@ -132,13 +136,36 @@ def __init__(
         self.known_future = self._check_known_future(known_future, df_exog)
         self._regressors = copy(self.known_future)
 
+        self.hierarchical_structure = hierarchical_structure
+        self.current_df_level: Optional[str] = self._get_dataframe_level(df=self.df)
+        self.current_df_exog_level: Optional[str] = None
+
         if df_exog is not None:
             self.df_exog = df_exog.copy(deep=True)
             self.df_exog.index = pd.to_datetime(self.df_exog.index)
-            self.df = self._merge_exog(self.df)
+            self.current_df_exog_level = self._get_dataframe_level(df=self.df_exog)
+            if self.current_df_level == self.current_df_exog_level:
+                self.df = self._merge_exog(self.df)
 
         self.transforms: Optional[Sequence["Transform"]] = None
 
+    def _get_dataframe_level(self, df: pd.DataFrame) -> Optional[str]:
+        """Return the level of the passed dataframe in hierarchical structure."""
+        if self.hierarchical_structure is None:
+            return None
+
+        df_segments = df.columns.get_level_values("segment").unique()
+        segment_levels = {self.hierarchical_structure.get_segment_level(segment=segment) for segment in df_segments}
+        if len(segment_levels) != 1:
+            raise ValueError("Segments in dataframe are from more than 1 hierarchical levels!")
+
+        df_level = segment_levels.pop()
+        level_segments = self.hierarchical_structure.get_level_segments(level_name=df_level)
+        if len(df_segments) != len(level_segments):
```

Comment:
I'm not sure if this check should be here or not.

Reply:
Why do you think so? You mean that it is not the real case?

Reply:
I don't really think that it should be in exactly this function. Check is useful, but it looks somewhat artificial in this function because we have already got that we wanted here.

Reply:
Ok, my problem with this function is what it checks this conditions every run. But according to the name it should just return the value and don't check correctness during each run.

Reply:
I'm generally ok with this, it is optional to fix.